### PR TITLE
Infer static type from the value itself for imported arrays/dictionaries

### PR DESCRIPTION
**Pull request by SupunS**
Work towards #870

## Description

Infer the static type from the array/dictionary value, if the expected type is not specific enough/ not available.

Depends on #1025 and #1048
______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

**Comment by SupunS** [comment]
The last two commits 1d26160 and f24e710 have the actual code changes. The rest is from the dependent PRs.

**Comment by turbolent** [comment]
@SupunS Do you think that we could hold off with this for now and get all the supertype inference PRs in later, i.e. the next release / spork? Do you think it's possible to do something simple for now, like e.g. infer `[AnyStruct]` for an array value where the expected static type is `AnyStruct`? (same for dictionaries)

**Comment by SupunS** [comment]
@turbolent Yeah, that should be fine. I added this more as a way to test the concept/usage.
I think arrays are rather simple. We can assume it to be `[AnyStruct]`
Same can be done for dictionary values. 

But for dictionary keys, `AnyStruct` might be a problem because it is not a valid type for keys. And might become a problem somewhere down the line?

File: runtime/convertValues.go [hunk 1]
```diff
@@ -663,8 +663,23 @@ func importArrayValue(
...
+			typ, err := inter.ConvertStaticToSemaType(value.StaticType())
+			if err != nil {
+				return nil, err
+			}
+			types[i] = typ
+		}
+
+		elementSuperType := sema.LeastCommonSuperType(types...)
+		if elementSuperType == sema.InvalidType {
+			return nil, fmt.Errorf("cannot import array: elements does not belong to the same type")
```
Comment (turbolent) [inline runtime/convertValues.go:678]:
```suggestion
			return nil, fmt.Errorf("cannot import array: elements do not belong to the same type")
```

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**